File: test_src/fortran/hierarchical_parallelism/atomic/double_precision/target__teams__parallel.f90
program target__teams__parallel

    USE OMP_LIB


    implicit none
  
    
    DOUBLE PRECISION :: COUNTER = 0

    
    INTEGER :: num_teams
    
    
    INTEGER :: num_threads
    

     
    
    !$OMP TARGET   MAP(TOFROM: COUNTER) 



    

    
    !$OMP TEAMS 



    
    num_teams = omp_get_num_teams()
    

    
    !$OMP PARALLEL 



    
    num_threads = omp_get_num_threads()
    

    

!$OMP ATOMIC UPDATE

counter = counter + 1./(num_teams*num_threads)


 
     

    !$OMP END PARALLEL
     

    !$OMP END TEAMS
     

    !$OMP END TARGET
    

    IF  ( ( ABS(COUNTER - 1) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected 1 Got', COUNTER
        call exit(1)
    ENDIF

end program target__teams__parallel

File: test_src/cpp/hierarchical_parallelism/reduction_atomic/double/target_teams__parallel.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#ifdef _OPENMP
#include <omp.h>
#else
int omp_get_num_teams() {return 1;}
int omp_get_num_threads() {return 1;}
#endif
bool almost_equal(double x, double gold, float tol) {
        return gold * (1-tol) <= x && x <= gold * ( 1+tol );
}
void test_target_teams__parallel(){
 // Input and Outputs
double counter{};
// Main program
#pragma omp target teams  map(tofrom:counter) 
{
const int num_teams = omp_get_num_teams();
 double partial_counter{};
#pragma omp parallel  reduction(+: counter)  
{
const int num_threads = omp_get_num_threads();
partial_counter += double { 1.0f/(num_teams*num_threads) } ;
}
#pragma omp atomic update
counter += partial_counter;
}
// Validation
if ( !almost_equal(counter,double { 1 }, 0.1)  ) {
    std::cerr << "Expected: " << 1 << " Got: " << counter << std::endl;
    throw std::runtime_error( "target_teams__parallel give incorect value when offloaded");
}
}
int main()
{
    test_target_teams__parallel();
}
